sql/api/src/main/scala/org/apache/spark/sql/expressions/ReduceAggregator.scala

```diff
@@ -27,10 +27,11 @@ import org.apache.spark.sql.{Encoder, Encoders}
  *
  * This class currently assumes there is at least one input row.
  */
+@SerialVersionUID(5066084382969966160L)
 private[sql] class ReduceAggregator[T: Encoder](func: (T, T) => T)
     extends Aggregator[T, (Boolean, T), T] {
 
-  @transient private val encoder = implicitly[Encoder[T]]
+  @transient private lazy val encoder = implicitly[Encoder[T]]
```

Comment:
Why is this needed?

Reply:
This transient encoder won't be serialised, so upon de-serialisation it'll be `null`. We now make it a `def`/`lazy val` so it can be re-calculated.

Reply:
We need this just now as we implement `reduceGroups` using this ReduceAggregator.